Cython/Compiler/TypeStubGenerator.py

```diff
@@ -0,0 +1,433 @@
+
+            if node.level > 0:
+                module_name = "%s%s" % ("." * node.level , module_name)
+
+            self.putline("from %s import %s" % (module_name, all_imported_children))
+
+        return node
+
+    # Optimized original code by having there be one function to take 
+    # the place of two of them I could see what Scoder meant when 
+    # said the original pull request needed to be cleaned up...
+
+    
+    def write_class(self, node, class_name):
+        self.endline()
+        self.put("class %s" % class_name)
+        if getattr(node,"bases",None) and isinstance(node.bases, TupleNode) and node.bases.args:
+            self.put("(")
+            self.put(",".join([name.name for name in node.bases.args]))
+            self.endline("):")
+        else:
+            self.endline(":")
+        self.class_func_count = 0
+        self._visitchildren_indented(node)
+        if self.class_func_count < 1:
+            self.indent()
+            self.putline("pass")
+            self.dedent()
+        self.class_func_count = 0
+        self.emptyline()
+        return node 
+    
+    # I have tried to merege these before via visit_ClassDefNode but it causes the system to break so this 
+    # was the best I could do to minigate the problem - Vizonex 
+    def visit_CClassDefNode(self, node):
+        return self.write_class(node, node.class_name)
+
+    def visit_PyClassDefNode(self, node):
+        return self.write_class(node, node.name)
+
+    def visit_CFuncDefNode(self, node):
+        # cdefs are for C only...
+        if not node.overridable:
+            return node 
+        
+        func_args = []
+        for arg in node.declarator.args:
+            value = ""
+            if not arg.declarator.name:
+                value = arg.base_type.name
+            elif hasattr(arg.base_type, "name"):
+                value = "%s : %s" % (arg.declarator.name, self.translate_base_type_to_py(arg.base_type))
+            if (arg.default is not None or
+                arg.default_value is not None):
+                value += " = ..."
+            func_args.append(value)
+        
+        self.class_func_count += 1
+        
+        func_name = node.declared_name()
+        self.startline()
+        self.put("def %s(" % func_name)
+    
+        self.put(", ".join(func_args))
+        
+        # TODO Maybe Try passing docstrings in the future for vscode users' sake
+        # or have it also be a compiler argument?...
+      
+        self.endline(") -> %s: ..." % self.translate_base_type_to_py(node.base_type))
+        
+        return node
+
+    
+    def write_decorator(self, decorator):
+        if isinstance(decorator, CallNode):
+            return
+        
+        self.startline("@")
+        if isinstance(decorator, NameNode):
+            self.endline("%s" % decorator.name)
+        else:
+            assert isinstance(decorator, AttributeNode) , "Decorator was not an attribute node..."
```

Comment:
I'd be worried how this interacts with https://peps.python.org/pep-0614/. There's definitely cases where decorators aren't attributes or names now.

I'm not quite sure how decorators mix with stub files though.

Reply:
Recursion again, if you want to properly serialise the stub files. But yeah, not sure if that's really intended here.